core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
OK, I tried making the whole method async, but that breaks a bunch of tests which makes me suspect that it will break some real behavior too. That brings me to the question that @lianhuiwang  asked above. Can we rely on the master's state? When is the master out of touch with the executor?

@andrewor14 - What do you think about avoiding the RPC and just replying on the state in the BMM?